### DISCUSSION TITLE
From DWG to (STEP, IGES, SAT or something else 3D)

### DISCUSSION THREAD

**Discussion by Mannshoch** [LibreDWG/libredwg#1028] [Q&A] [answered]
I't is not required to have any other Information than the 3D-Data.

And does it also work with Proxy object?

Currently I have access to AutoCAD but even If I export to DXF (2000 or 2018) or IGES , The STEP seems not possible to export (no proper body found) and the DXF seems not readable eg for FreeCAD.

At moment the only way to get 3D models is with Inventor. 

What is the best way to get all 3D-Objects (Mesh, brep, swept-solid, CSG) into a neutral file?

**Comment by Mannshoch** [comment]
Could I create some TestData that may be usefull for you? I have access to autoCAD and Inventor.

**Comment by rurban** [comment]
Sure. I don't have access to AutoCAD nor Inventor

**Reply by Mannshoch** [reply]
For Inkscape I created these files: https://gitlab.com/inkscape/inbox/-/issues/2076#note_309337210 (at that time I had only access to AutoCAD LT)
If you wish I could create something like that for you. I'm not a developer but if you give me details, example or other kind of advice I will create sample files for you.

**Reply by Mannshoch** [reply]
I created a set of data (xref) which are converted to different DWG versions.
[Test-Data.zip](https://github.com/user-attachments/files/18410314/Test-Data.zip)

**Reply by michal-josef-spacek** [reply]
@Mannshoch What is a difference between zip files? (Level0 - 2000.zip vs Level0 - 2004.zip)

If I am looking to it I see:
* difference in file size of Level0.dwg, Stair.dwg, TESTDATA - 2018.dwg, Pipe.dwg and Trichter2.dwg
* All DWG versions are AC1032

**Reply by Mannshoch** [reply]
AutoCAD gives the Possibility to convert a DWG into other version.
https://www.autodesk.com/support/technical/article/caas/sfdcarticles/sfdcarticles/How-to-convert-a-dwg-file-to-a-previous-format.html

I used this Convert tool to create DWG Files for all AUTOCAD DWG versions.

**Reply by michal-josef-spacek** [reply]
@Mannshoch I know, but these files are all AC1032 (latest version). The string at the beginning of binary data in the file. You could look at it simply.

**Reply by Mannshoch** [reply]
weird, AutoCAD 2025 seems not Converting. None of the settings I changed created a converted file. An older AutoCAD seems working. I created the Converted files again.
[Test-Data2.zip](https://github.com/user-attachments/files/18440192/Test-Data2.zip)

**Reply by michal-josef-spacek** [reply]
ok, fine :-) Thank you.

**Reply by Mannshoch** [reply]
Where the Test files helpfull?
Could I do anything more to help you? 
On my current test with libredwg-0.13.3.7825-win64 to convert dwg to dxf I still got a dxf without bodies.

```
ERROR: Invalid SAB len=4 [RC]
ERROR: Invalid SAB len=4 [RC]
ERROR: Invalid SAB len=4 [RC]
ERROR: Invalid SAB len=4 [RC]
ERROR: Invalid SAB len=4 [RC]
ERROR: Invalid SAB len=4 [RC]
ERROR: Invalid SAB len=4 [RC]
ERROR: Invalid SAB len=4 [RC]
ERROR: Invalid SAB len=4 [RC]
ERROR: Invalid SAB len=4 [RC]
ERROR: Invalid SAB len=4 [RC]
ERROR: Invalid SAB len=4 [RC]
Warning: Unstable Class object 501 MATERIAL (0x481) 14/11
Warning: Unstable Class object 501 MATERIAL (0x481) 22/19
Warning: Unstable Class object 501 MATERIAL (0x481) 30/21
Warning: Unhandled Object TABLESTYLE in out_dxf 95/65
Warning: Unstable Class object 505 MLEADERSTYLE (0xfff) 97/67
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unstable Class object 513 ACAD_EVALUATION_GRAPH (0x481) 137/1D8
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unstable Class object 513 ACAD_EVALUATION_GRAPH (0x481) 141/1DE
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unstable Class object 513 ACAD_EVALUATION_GRAPH (0x481) 172/452
Warning: Unstable Class object 513 ACAD_EVALUATION_GRAPH (0x481) 176/472
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unstable Class object 513 ACAD_EVALUATION_GRAPH (0x481) 180/48F
Warning: Unstable Class object 513 ACAD_EVALUATION_GRAPH (0x481) 191/4FC
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unstable Class object 513 ACAD_EVALUATION_GRAPH (0x481) 195/504
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unstable Class object 513 ACAD_EVALUATION_GRAPH (0x481) 203/6EC
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unstable Class object 513 ACAD_EVALUATION_GRAPH (0x481) 211/6F5
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unknown object, skipping eed/reactors/xdic
Warning: Unstable Class object 513 ACAD_EVALUATION_GRAPH (0x481) 226/8CB
Warning: Unknown object, skipping eed/reactors/xdic
```

**Reply by michal-josef-spacek** [reply]
Test files are helpful, but the number of issues is big.
e.g. I created https://github.com/LibreDWG/libredwg/issues/1177 found in your test files.

**Reply by michal-josef-spacek** [reply]
Previous one is fixed.
I created another one: https://github.com/LibreDWG/libredwg/issues/1178
:-)

**Reply by michal-josef-spacek** [reply]
The previous one is fixed.
I created another one: https://github.com/LibreDWG/libredwg/issues/1179

**Comment by rurban** [comment, marked as answer]
No, we cannot decode those undocumented classes: (from Level0)

Warning: Unknown Class object 0 ACAMGHOLECHARTSTDDIN (0x62)
Warning: Unknown Class object 0 ACMDATAENTRYBLOCK (0x420a)
Warning: Unknown Class object 0 ACMTHREADLINESTANDARDDIN (0x420a)
Warning: Unknown Class object 0 ACMWELDSTANDARDDIN2014 (0x420a)
Warning: Unknown Class object 0 AMCONTEXTMGR (0x420a)
Warning: Unknown Class object 0 MAPIAMMOUDLE (0x420a)
Warning: Unknown Class object 101 ACMSECTIONSTANDARDCUSTOM (0x98)
Warning: Unknown Class object 101 AMDTNOTETEMPLATEDB (0xcc)
Warning: Unknown Class object 128 ACGREFACADMASTER (0x40is_zombie)
Warning: Unknown Class object 128 AMGSETTINGSOBJ (0x0is_zombie)
Warning: Unknown Class object 16906 AMIMASTER (0x28e)
Warning: Unknown Class object 16 ACIMSYSVARMAN (0xf320)
Warning: Unknown Class object 16 ACMANOOTATIONVIEWSTANDARDDIN (0x3d28)
Warning: Unknown Class object 261 ACMTAPERSTANDARDDIN (0x19)
Warning: Unknown Class object 32823 ACAMGTITLESTD (0xdcis_zombie)
Warning: Unknown Class object 34960 ACAMCOMPDEF (0x100is_zombie)
Warning: Unknown Class object 34960 ACAMCOMPDEFMGR (0x100is_zombie)
Warning: Unknown Class object 34960 ACAMGRCOMPDEF (0x100is_zombie)
Warning: Unknown Class object 34960 ACAMPROJECT (0x100is_zombie)
Warning: Unknown Class object 34 ACMDETAILTANDARDCUSTOM (0x92is_zombie)
Warning: Unknown Class object 4096 ACAMGDIMSTD (0x8600)
Warning: Unknown Class object 49212 ACIMINTSYSVAR (0x40)
Warning: Unknown Class object 5120 ACAMCONTEXTMODELER (0x44e)
Warning: Unknown Class object 5120 ACIMSTRSYSVAR (0x440)
Warning: Unknown Class object 5120 ACMIDSTANDARDDIN2012 (0x466)
Warning: Unknown Class object 5120 ACMSURFSTANDARDDIN2002 (0x460)
Warning: Unknown Class object 5908 ACMSTANDARDDIN (0x4)
Warning: Unknown Class object 6676 ACMBOMSTANDARDDIN (0x0)
Warning: Unknown Class object 6784 ACAMGLAYSTD (0x0)
Warning: Unknown Class object 6932 ACMSECTIONSTANDARDDIN (0x0)
Warning: Unknown Class object 7444 ACMNOTESTANDARDDIN (0x4004)
Warning: Unknown Class object 8416 ACMCENTERLINESTANDARDDIN (0x5000)
Warning: Unknown Class object 8416 ACMDATADICTIONARY (0x5000)
Warning: Unknown Class object 8416 ACMDATUMSTANDARDDIN2008 (0x5000)
Warning: Unknown Class object 8416 ACMDETAILSTANDARDDIN (0x5000)
Warning: Unknown Class object 8416 ACMFCFSTANDARDDIN2008 (0x5000)
Warning: Unknown Class object 8416 AMDTEDGESTANDARDDIN13715 (0x5000)
Warning: Unknown Class object 8 ACGREFMASTER (0x7910is_zombie)
Warning: Unknown Class object 8 AMGOBJPROPCFG (0x0)

Level1 has these:
Warning: Unknown Class object 0 ACAMCOMPDEF (0x0)
Warning: Unknown Class object 0 ACAMCOMPDEFMGR (0xf02)
Warning: Unknown Class object 0 ACAMGRCOMPDEF (0x4)
Warning: Unknown Class object 0 ACGREFACADMASTER (0x420a)
Warning: Unknown Class object 0 ACMDATADICTIONARY (0x420a)
Warning: Unknown Class object 0 ACMDETAILTANDARDCUSTOM (0x1001)
Warning: Unknown Class object 0 AMDTNOTETEMPLATEDB (0x6003)
Warning: Unknown Class object 0 AMIMASTER (0x420a)
Warning: Unknown Class object 12672 ACAMGHOLECHARTSTDDIN (0x0)
Warning: Unknown Class object 129 ACAMPROJECT (0x200)
Warning: Unknown Class object 129 ACMNOTESTANDARDDIN (0x240)
Warning: Unknown Class object 16906 AMGSETTINGSOBJ (0x28e)
Warning: Unknown Class object 17 ACIMSYSVARMAN (0x100)
Warning: Unknown Class object 1 ACMTHREADLINESTANDARDDIN (0x200)
Warning: Unknown Class object 204 ACMSECTIONSTANDARDCUSTOM (0x100)
Warning: Unknown Class object 256 AMCONTEXTMGR (0x0)
Warning: Unknown Class object 256 AMGOBJPROPCFG (0x0)
Warning: Unknown Class object 32935 ACAMGLAYSTD (0xc4is_zombie)
Warning: Unknown Class object 33536 ACMANOOTATIONVIEWSTANDARDDIN (0x100)
Warning: Unknown Class object 33536 ACMTAPERSTANDARDDIN (0x100)
Warning: Unknown Class object 4372 ACGREFMASTER (0x4)
Warning: Unknown Class object 5120 ACMIDSTANDARDDIN2012 (0x478)
Warning: Unknown Class object 5120 ACMSECTIONSTANDARDDIN (0x480)
Warning: Unknown Class object 5140 MAPIAMMOUDLE (0x4)
Warning: Unknown Class object 6420 ACMDATAENTRYBLOCK (0x4)
Warning: Unknown Class object 7188 AMDTEDGESTANDARDDIN13715 (0x4)
Warning: Unknown Class object 7444 ACMFCFSTANDARDDIN2008 (0x4)
Warning: Unknown Class object 7720 ACIMINTSYSVAR (0x8)
Warning: Unknown Class object 7956 ACMBOMSTANDARDDIN (0x4)
Warning: Unknown Class object 8416 ACAMCONTEXTMODELER (0x5000)
Warning: Unknown Class object 8416 ACIMSTRSYSVAR (0x5000)
Warning: Unknown Class object 8416 ACMCENTERLINESTANDARDDIN (0x5000is_zombie)
Warning: Unknown Class object 8416 ACMDATUMSTANDARDDIN2008 (0x5000)
Warning: Unknown Class object 8416 ACMDETAILSTANDARDDIN (0x5000)
Warning: Unknown Class object 8416 ACMSTANDARDDIN (0x5000)
Warning: Unknown Class object 8416 ACMSURFSTANDARDDIN2002 (0x5000)
Warning: Unknown Class object 8416 ACMWELDSTANDARDDIN2014 (0x5000)
Warning: Unknown Class object 88 ACAMGDIMSTD (0x180)
Warning: Unknown Class object 97 ACAMGTITLESTD (0xc8)

**Reply by Mannshoch** [reply]
Is that cause because of encrypted data or has it other causes. Could I as CAD user do anything for you about it?

**Reply by michal-josef-spacek** [reply]
@Mannshoch I think that these objects are Inventor objects. Do the Inventor export to DXF? A DWG file can contain any entity or object. e.g. You could create your own entity/object.

Or export to DXF by AutoCAD. I need to look at DWG and DXF together (ideally from the same software).

**Reply by Mannshoch** [reply]
I did not open and saved it with Inventor. Level 0 Itself should not contain any model data. It contains only xrefs to other dwg files. It is may possible that these data contain some cached data to show a body. I know these concept from Solidworks. Where every assembly contain the tessellated graphic data for it's sub-assembly and sub-parts. It allows that you could share only the assembly file if you only need a graphical representation in the Solidworks viewer and also the file could be loaded faster and show you something even not all files are loaded.

**Comment by rurban** [comment]
Some usage examples:

`jq '.OBJECTS[] | select(.acis_data != null) | .acis_data' example_2000.json`

or `dwgfilter '.OBJECTS[] | select(.acis_data != null) | .acis_data' example_2000.dwg`

this prints all SAT and SAB objects. But not any custom objects, because these are not spec'ed and hidden on purpose.
